Sources/SwiftBasicFormat/BasicFormat.swift

```diff
@@ -171,7 +171,50 @@ open class BasicFormat: SyntaxRewriter {
       }
     }
 
-    return false
+    let syntaxToken: Syntax = Syntax(token)
+    switch syntaxToken.keyPathInParent {
+    case \StringLiteralExprSyntax.closeQuote:
+      return token.tokenKind == .multilineStringQuote
+    default:
+      return false
+    }
+  }
+
+  open func requiresNewline(between first: TokenSyntax?, and second: TokenSyntax?) -> Bool {
+      switch (first?.tokenKind, second?.tokenKind) {
+      case (.atSign, _),
+          (.colon, _),
+          (.comma, _),
+          (.eof, _),
+          (.equal, _),
+          (.identifier, .comma),
+          (.identifier, .leftBrace),
+          (.identifier, .leftParen),
+          (.identifier, .period),
+          (.identifier, .rightParen),
+          (.keyword, .arrow),
+          (.keyword(.await), .identifier),
+          (.keyword(.async), _),
+          (.keyword(.false), _),
+          (.keyword(.func), _),
+          (.keyword(.let), .identifier),
+          (.keyword(.let), .wildcard),
+          (.keyword(.throws), .rightBrace),
+          (.keyword(.true), _),
+          (.keyword(.var), .identifier),
+          (.keyword(.var), .wildcard),
+          (.leftParen, _),
+          (.period, .identifier),
+          (.rightParen, .identifier),
+          (.rightParen, .keyword(.async)),
+          (.rightSquareBracket, _),
+          (_, .colon),
+          (_, .eof),
+          (_, .equal):
+          return false
+      default:
+          return true
+      }
```

Comment:
@ahoppen did you think on something like this?

Note: This is not the final result. Did just try to pass the first couple of test cases. 
Not sure if we should default to false and then only return true if needed. 

And if we should try to get the previous code from `requireLeadingNewline` into this method or not.

Reply:
I was thinking more of something like this that’s very much based on `requiresLeadingNewline` and actually replaces `requiresLeadingNewline`. 

```swift
open func requiresNewline(between first: TokenSyntax?, and second: TokenSyntax?) -> Bool {
  // NOTE: From requiresLeadingNewline

  // We don't want to add newlines inside string interpolation
  if isInsideStringInterpolation(first) && token.tokenKind != .multilineStringQuote {
    return false
  }

  // NOTE: From AnyKeyPath.requiresLeadingTrivia, which would become obsolete
  switch second.keyPathInParent {
  case \AccessorBlockSyntax.rightBrace:
    return true
  case \ClosureExprSyntax.rightBrace:
    return true
  case \CodeBlockSyntax.rightBrace:
    return true
  case \IfConfigClauseSyntax.poundKeyword:
    return true
  case \IfConfigDeclSyntax.poundEndif:
    return true
  case \MemberDeclBlockSyntax.rightBrace:
    return true
  case \SwitchExprSyntax.rightBrace:
    return true
  default:
    break
  }

  // NOTE: Whatever you need to add for your PR
  switch first.keyPathInParent {
  case \StringLiteralExprSyntax.openQuote where first.tokenKind == .multilineStringQuote:
    return true
  default:
    break
  }

  // NOTE: From requiresLeadingNewline
  var ancestor: Syntax = Syntax(second)
  while let parent = ancestor.parent {
    ancestor = parent
    if ancestor.position != second.position || ancestor.firstToken(viewMode: viewMode) != second {
      break
    }
    if let ancestorsParent = ancestor.parent, childrenSeparatedByNewline(ancestorsParent) {
      return true
    }
    switch ancestor.keyPathInParent {
    case \IfConfigClauseSyntax.elements:
      return true
    default:
      break
    }
  }

  return false
}
```